### ISSUE TITLE
The "Remove Log" Can Be More Efficient (Leaf Pos Only)

### ISSUE THREAD

**Issue by antiochp** [mimblewimble/grin#1137]
Right now we store all removed positions in the rm_log. This includes both leaf pos (the actual entries in the MMR) and non-leaf parent positions. As we gradually remove subtrees all the non-leaf parent positions are added to the rm_log.

So for example in a very simple MMR with 2 elements added (in pos 1 and pos 2, parent at pos 3) - 
```
 3
 /\
1  2
```

If we remove pos 1, then the rm_log will contain `[1]`.
\* technically it will contain `[(1, 1)]` as we store the "index" alongside each removed position, but ignoring this for now.
If we remove pos 1 and pos 2, we can also remove pos 3 and the rm_log will look like `[1, 2, 3]` etc.

I would like to argue that storing non-leaf positions in the remove log is both - 
  a) expensive in terms of storage requirements (specifically the number of non-leaf nodes in the rm_log approaches the number of leaf nodes over time).
  b) expensive in terms of searches in the rm_log for "has this pos been removed?" behavior

During regular gossip we care about "has this pos been removed?" __only__ for leaf positions, for both block and transaction validation and processing. Having non-leaf positions in the rm_log only makes the searches more expensive and inflates the size of the rm_log on disk (and in memory).

The _only_ time (I believe) we care about non-leaf positions is during `check_compact` when we prune and compact the MMR to correctly offset the compacted hash and data files.
But I think it is reasonably straightforward to build an "expanded" rm_log based on a "leaf only" rm_log and an existing prunelist.
In the trivial case above we want to simply convert `[1, 2]` into the expanded `[1, 2, 3]`.

If we do this I think there is a nice clear relationship between the rm_log and the prunelist -
* rm_log: leaf pos only
* prunelist: roots of subtrees only

And everything else can be reconstructed from the interaction of the two.

I know we had discussed possibly making the rm_log behave more like the prunelist, where we stored roots of subtrees in the rm_log. But I'm not convinced we gain anything from doing that.
It would make it more expensive to determine if leaves were present in the rm_log.
And I'm not sure it would be any more compact as we would still need to account for rewind and owuld not be able to roll up multiple remove operations over time into a single subtree etc.

Working on getting a proof-of-concept up for this to see what kind of space savings we actually get from a leaf-only rm_log.
And more importantly to see if I can actually get this to work...

**Comment by ignopeverell** [comment]
Totally agree the rm log is suboptimal. The prune list might be as well. I haven't had time to think of what the most optimal data structure would be for this, so thanks for looking at it.

**Comment by antiochp** [comment]
Another thing that might be worth looking at (think we've discussed this before) is representing the rm_log as a bitmap/bitset.
Just need to think about how rewind might work (and how to track the changes per index etc.)

Found this from some searching - https://roaringbitmap.org/
And the "not ready for production use" rust wrapper - https://github.com/saulius/croaring-rs


I do think we could easily get the rm_log to roughly half its current size by skipping the parent nodes in there though.
I'm not sure the prunelist could be any more efficient though - we only store the roots of pruned subtrees. Haven't thought too much about that.

**Comment by ignopeverell** [comment]
We can't do more than halve the rm log, given that for a MMR of N leaves, there's always less than N parents (a full binary tree would have N-1). But even reducing its size by 30% would be good and benefit lookup time.

A compressed bitmap could be nice as a replacement of both rm log and prune list with a single array of bits that just indicate what has been spent or not. Merging the rm log would just be a `and` operation. But as you mention rewind and calculating the data file offset could be tricky...

**Comment by antiochp** [comment]
Yeah just under 50% reduction is the "best case" as there are always more leaves than parents.

Got something running locally just storing leaves in the rm_log.
These are number from a couple of full archive syncs - 

Current rm_log impl - 
```
# .grin/txhashset/output/
-rw-r--r-- 1 4306984 Jun  4 22:32 pmmr_data.bin
-rw-r--r-- 1 8106912 Jun  4 22:32 pmmr_hash.bin
-rw-r--r-- 1  824172 Jun  4 22:32 pmmr_rm_log.bin
```

Leaf only rm_log impl -
```
# .grin/txhashset/output/
-rw-r--r-- 1 4306984 Jun  4 23:33 pmmr_data.bin
-rw-r--r-- 1 8106912 Jun  4 23:33 pmmr_hash.bin
-rw-r--r-- 1  498324 Jun  4 23:33 pmmr_rm_log.bin
```

The rm_log is almost 40% smaller. 
Kind of surprised we see so much reduction on testnet2 actually...

**Comment by antiochp** [comment]
Hmm - using a single bitmap for _both_ the rm_log and the prunelist is interesting.
The pos never conflict (one cares about leaves, the other cares strictly about roots of subtrees) so we can just overlay them both in the same bitmap.

I wonder if we can handle rewind by simply "reverting" individual blocks.
We need full blocks to be able to rewind, so we know all the outputs and their corresponding pos.
I'd guess that most rewinds happen for only the one or two most recent blocks.
Assuming only rarely would we encounter a rewind for an extended number of blocks (but we'd still need to handle this efficiently). 
And even so we can still only do this safely within the horizon for pruned nodes.

**Comment by antiochp** [comment]
Resolved in #1147.
